### PR TITLE
Handle References on Serialization

### PR DESCRIPTION
**Pull request by jozkee**
Contributes to #41002

Exposes new API `JsonSerializerOptions.ReferenceHandlingOnSerialize` along with the `ReferenceHandlingOnSerialize` enum.

The available options are:
* Error: throws a `JsonException` when a circular reference is detected in the object's branch (detection starts when MaxDepth is reached).
* Ignore: Instead of throwing, ignores properties/objects that cause circularity in the object's branch.
* Preserve: Writes `"$id": "#"` properties in the JSON complex types that can be later use with the `$ref` property to denote reference, these references are not limited to the object's branch but they live in the entire object graph.

### CODE REVIEW

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your corefx repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/corefx <path to the patch created in step 1>`

File: src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleReference.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+            JsonReferenceHandlingAttribute attr = JsonPropertyInfo.GetAttribute<JsonReferenceHandlingAttribute>(state.Current.JsonPropertyInfo?.PropertyInfo);
+            if (attr != null)
+            {
+                handling = attr.Handling;
+            }
+            //Class has attribute?
+            else if ((attr = state.Current.JsonClassInfo.Type.GetCustomAttribute<JsonReferenceHandlingAttribute>()) != null)
+            {
+                handling = attr.Handling;
+            }
```
Comment (jozkee) [inline src/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Write.HandleReference.cs]: We might consider dropping attribute support... this is causing a huge performance decrease on the main path (around 80% increase in runtime).

@ahsonkhan @steveharter thoughts or suggestion on how to handle this better?

**Comment by jozkee** [comment]
Performance results.

Before:

|                               Type |            Method |         Mean |        Error |       StdDev |       Median |          Min |          Max |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|----------------------------------- |------------------ |-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|--------:|--------:|--------:|----------:|
|          WriteJson\<IndexViewModel\> | SerializeToString |  24,054.7 ns |    249.37 ns |    233.26 ns |  24,096.0 ns |  23,715.0 ns |  24,427.6 ns |  6.0730 |       - |       - |   25608 B |
|                WriteJson\<Location\> | SerializeToString |   1,141.8 ns |     22.11 ns |     21.72 ns |   1,144.2 ns |   1,086.7 ns |   1,174.1 ns |  0.1392 |       - |       - |     584 B |
|          WriteJson\<LoginViewModel\> | SerializeToString |     461.3 ns |      9.24 ns |     14.39 ns |     454.5 ns |     435.8 ns |     500.6 ns |  0.0820 |       - |       - |     344 B |
| WriteJson\<MyEventsListerViewModel\> | SerializeToString | 581,376.1 ns | 11,130.34 ns | 14,472.59 ns | 578,265.6 ns | 560,069.5 ns | 612,906.1 ns | 46.8750 | 46.8750 | 46.8750 |  275944 B |


After:

|                               Type |            Method |         Mean |        Error |       StdDev |       Median |          Min |          Max |   Gen 0 |   Gen 1 |   Gen 2 | Allocated |
|----------------------------------- |------------------ |-------------:|-------------:|-------------:|-------------:|-------------:|-------------:|--------:|--------:|--------:|----------:|
|          WriteJson\<IndexViewModel\> | SerializeToString |  25,104.2 ns |    295.41 ns |    276.33 ns |  24,979.4 ns |  24,686.5 ns |  25,627.6 ns |  6.0730 |       - |       - |   25568 B |
|                WriteJson\<Location\> | SerializeToString |   1,119.0 ns |     21.16 ns |     19.79 ns |   1,117.9 ns |   1,085.5 ns |   1,147.8 ns |  0.1392 |       - |       - |     584 B |
|          WriteJson\<LoginViewModel\> | SerializeToString |     467.9 ns |      9.17 ns |     14.28 ns |     464.7 ns |     444.3 ns |     496.9 ns |  0.0811 |       - |       - |     344 B |
| WriteJson\<MyEventsListerViewModel\> | SerializeToString | 563,753.3 ns | 11,171.98 ns | 16,022.52 ns | 562,761.5 ns | 543,714.4 ns | 617,808.5 ns | 46.8750 | 46.8750 | 46.8750 |  276440 B |

File: src/System.Text.Json/ref/System.Text.Json.cs [hunk 2]
```diff
@@ -84,6 +84,14 @@ public enum JsonCommentHandling : byte
         Skip = (byte)1,
         Allow = (byte)2,
     }
+
```
Comment (ahsonkhan) [inline src/System.Text.Json/ref/System.Text.Json.cs:87]: nit: make sure to auto-generate the ref (i.e. remove the extra new lines).
https://github.com/dotnet/corefx/blob/master/Documentation/coding-guidelines/updating-ref-source.md

**Comment by maryamariyan** [comment]
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.

**Pull request closed without merging**